keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,852 @@
+
+```yaml
+rules:
+  - operations: ['CREATE']
+    apiGroups: '' # core
+    apiVersions: '*'
+    resources: 'pods'
+matchConditions:
+  # Only include pods with an NFS volume.
+  - expression: 'object.spec.volumes.exists(v, v.has(nfs))'
+```
+
+### Goals
+
+1. Provide a filtering mechanism for excluding requests from an admission webhook
+2. Maintain consistency with `ValidatingAdmissionPolicy`
+
+### Non-Goals
+
+* Provide a mechanism to exclude requests from all webhooks.
+
+## Proposal
+
+### API
+
+Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
+with a new `MatchConditions` field:
+
+```go
+
+type ValidatingWebhook struct {
+  // ...
+
+  // MatchConditions is a list of conditions on the AdmissionRequest ('request') that must be met
+  // for a request to be sent to this webhook. All conditions in the list must evaluate to TRUE for
+  // the request to be matched.
+  // +optional
+  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
+}
+
+type MutatingWebhook struct {
+  // ...
+  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
+}
+
+// MatchCondition represents a condition which must by fulfilled for a request to be sent to a webhook.
+type MatchCondition struct {
+  // NOTE: Placeholder documentation, to be replaced by https://github.com/kubernetes/website/issues/39089.
+  //
+	// Expression represents the expression which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// CEL expressions have access to the contents of the AdmissionRequest, organized into CEL variables:
+	//
+	// 'object' - The object from the incoming request. The value is null for DELETE requests.
+	// 'oldObject' - The existing object. The value is null for CREATE requests.
+	// 'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	//
+	// Required.
+	Expression string `json:"expression"`
+}
+```
+
+The match condition expression is evaluated by the same libraries as those used for CEL
+ValidatingAdmissionPolicy. The only difference in expressions is the availability of the `params`
+variable. Expressions requiring access to additional information outside the AdmissionRequest must
+be performed in the webhook, and are out of scope for this proposal.
+
+### Risks and Mitigations
+
+#### Security
+
+**Risk: Attacker adds or changes a match condition to weaken an admission policy.**
+
+This is does not represent a new threat, as doing so would require update access to the admission
+registration object, and with that permission an attacker could already disable the policy through
+manipulating match rules, namespace selector, or object selector (or reroute the webhook entirely).
+
+**Risk: Logic error in match condition expression.**
+
+Currently the match conditions must be encoded in the webhook backend itself. Moving the logic
+into a CEL expression does not materially increase the risk of a logic bug.
```

Comment:
Can you unit-test CEL expressions against input yamls? I'd argue that ease of testing does materially decrease the risk of a logic bug.

Reply:
Good point. I think the cell ecosystem is currently missing a lot of tools, so this is something that could be made a lot easier. I'll add a comment about it.